Build Multi-Arch Images

oliver-goetz opened the pull request:
**What this PR does / why we need it**:
This PR enables the CI pipeline to publish multi-arch images with support for linux/amd64 and linux/arm64.

**Which issue(s) this PR fixes**:
Fixes parts of https://github.com/gardener/gardener/issues/6258

**Special notes for your reviewer**:
I created a new `Dockerfile` for `cluster-autoscaler` which is easier to use in `docker-buildx` environment.

**Release note**:
<!--
Write your release note:
1. Enter your release note in the below block.
2. If no release note is required, just write "NONE" within the block.

Format of block header: <category> <target_group>
Possible values:
- category:       breaking|feature|bugfix|doc|other
- target_group:   user|operator|developer|dependency
-->
```other developer
Docker images for cluster-autoscaler are published with multi-arch support for `linux/amd64` and `linux/arm64` now.
Update Go version to `1.19.5`.
```

oliver-goetz committed:
Build Multi-Arch Images

oliver-goetz commented:
The failing `test-and-verify` check seems to be unrelated to my changes.

unmarshall reviewed (commented):
Looks good

himanshu-kun reviewed (commented):
I am assuming the variable `TARGETARCH` is set when using docker-buildx ?
Asking this because for similar PR in MCM (https://github.com/gardener/machine-controller-manager/pull/732/files) this was not there
Could you explain why it is different here?

oliver-goetz commented:
At `machine-controller-manager` the image build [runs in a Docker build image](https://github.com/gardener/machine-controller-manager/blob/7ec9a4261f4a7edf5b7e625b85dbc87375233f92/Dockerfile#L7). Thus, docker-buildx will use different build images depending on the target architecture and compiles the binary natively in its environment.

Here the images are cross-compiled in the [build step](https://github.com/gardener/autoscaler/blob/61b4a9600b538aaeaed3d1fdf6ed80c0f3c5266d/.ci/pipeline_definitions#L29-L31) and docker-buildx just copies these binaries into the right target image.

I was not aware of the the MCM build. Now that I know it, I think it might make sense using the same way for `cluster-autoscaler` too. WDYT?

himanshu-kun commented:
Yes I was also thinking the same, we should keep CA build process same as MCM.

timuthy commented on cluster-autoscaler/Dockerfile:
```diff
@@ -0,0 +1,8 @@
+
```
In general, I suggest to follow the best practice from the other projects in the `gardener` org and build the binaries directly in the container and use a multi-stage build in this file.

timuthy reviewed (commented):
Thanks for the PR, please see my comment in line.

oliver-goetz committed:
Switch to using a build image for cluster-autoscaler

oliver-goetz commented:
I switched to using a build image for cluster-autoscaler now.
Additionally, I introduced two other harmonizations.
- `cluster-autoscaler` is built using modules now
- Go version is updated to 1.19.4 as the go.mod file of cluster-autoscaler suggests 

The `build_oci_image_cluster-autoscaler` step seems to fail because it is still using the previous pipeline definition, but the file name of the cluster-autoscaler binary changed (no more architecture suffix). Please double-check this.

Btw. I noticed that using a build image increases the build time of the arm64 binary from 3,5 min to almost 15 min on my machine. I think this is because buildx creates a virtual arm64 environment on its native amd64 machine which impacts the performance. That might not be an issue, because we do not release that often, but I wanted to mention it anyway.

timuthy reviewed (approved):
/lgtm
Thank you!

oliver-goetz commented:
There are different versions of `cluster-autoscaler` in use in gardener/gardener starting with v1.20. 
Shall we cherry-pick this for the respective `rel-v1.*` branches too? Btw. there are no release branches for v1.24 and v1.25 yet.

himanshu-kun reviewed (changes requested):
could we have the `Dockerfile` moved to `cluster-autoscaler/Dockerfile` as we have 3 components under autoscaler repo, but the Dockerfile is just for the `cluster-autoscaler` component

himanshu-kun commented:
> The build_oci_image_cluster-autoscaler step seems to fail because it is still using the previous pipeline definition, but the file name of the cluster-autoscaler binary changed (no more architecture suffix). Please double-check this.

Confirmed.

> There are different versions of cluster-autoscaler in use in gardener/gardener starting with v1.20.
Shall we cherry-pick this for the respective rel-v1.* branches too? Btw. there are no release branches for v1.24 and v1.25 yet.

Yes we'll need that. I have created 1.24 and 1.25 branch now.

oliver-goetz committed:
Move Dockerfile into cluster-autoscaler directory

oliver-goetz commented:
> could we have the `Dockerfile` moved to `cluster-autoscaler/Dockerfile` as we have 3 components under autoscaler repo, but the Dockerfile is just for the `cluster-autoscaler` component

Docker needs to run in the root directory of this repo that we can copy `.ci/build` script into the build container. That's why I put the `Dockerfile` into the root directory too.
I moved it back to `cluster-autoscaler` directory. Hopefully the publish step of the pipeline can handle this situation too..

oliver-goetz committed:
Update to Go 1.19.5

himanshu-kun commented on .ci/pipeline_definitions:
```diff
@@ -17,13 +21,12 @@ autoscaler:
             image: 'eu.gcr.io/gardener-project/gardener/autoscaler/cluster-autoscaler'
-            dockerfile: 'Dockerfile.amd64'
-            dir: 'cluster-autoscaler'
+            dockerfile: './cluster-autoscaler/Dockerfile'
```
If you write it this way (this is how it was written earlier also when dockerfile was in `cluster-autoscaler` directory) https://github.com/gardener/autoscaler/blob/4d7787f19559533c5f2e9a0aef710b9000e84523/.ci/pipeline_definitions#L20-L21
Then I think we will be confident abt the publish step

oliver-goetz replied:
Then docker would not be able to copy the `.ci/build` script into the build container. It does not allow copying files from parent directories into containers.

himanshu-kun replied:
oh okay, then I think no changes required.

himanshu-kun reviewed (changes requested):
Thanks for the changes , asked to update the pipeline_definitions

himanshu-kun reviewed (approved):
/lgtm
